File: leetcode/2218.从栈中取出 K 个硬币的最大面值和.cpp
class Solution {
public:
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        int n = piles.size();
        std::vector<std::vector<int>> pre(n, std::vector<int>());
        
        // 预处理前缀和
        for (int i = 0; i < n; ++i) {
            pre[i].push_back(0);
            for (int x : piles[i]) {
                pre[i].push_back(pre[i].back() + x);
            }
        }

        std::vector<std::vector<int>> f(n + 1, std::vector<int>(k + 1, 0));

        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= k; ++j) {
                // 不选
                f[i][j] = f[i - 1][j];

                // 选
                for (int t = 1; t <= std::min(static_cast<int>(piles[i - 1].size()), j); ++t) {
                    f[i][j] = std::max(f[i][j], f[i - 1][j - t] + pre[i - 1][t]);
                }
            }
        }

        return f[n][k];
    }
};

File: leetcode/2218.从栈中取出 K 个硬币的最大面值和.py
class Solution:
    def maxValueOfCoins(self, p: List[List[int]], k: int) -> int:
        # 动态规划问题
        # 多重背包模板题
        # 就是对于每一个桶，需要分配几个空间，使得最后的结果最大
        # f[i][j]表示前i个桶分配j个空间的最大值是多少
        # 选或者不选，不选就是前i-1，选就是第i个桶分配几个的问题
        # f[i][j] = max(f[i - 1][j], f[i - 1][j - 1] + sum(piles[i][:1]), f[i - 1][j - 2] + sum(piles[i][:2], f[i-1][j-j] + sum(piles[:j]))
        # f[i][j - 1] = max(f[i - 1][j - 1], f[i - 1][j - 2] + sum(piles[i][:1]))
        # 状态个数 2000 * 1000 = 2e6
        # 预处理前缀和
        f = [0] * (k + 1)
        for a in p:
            g = f[:]
            s = 0
            for i in range(len(a)):
                s += a[i]
                for j in range(i + 1, k + 1):
                    g[j] = max(g[j], f[j - (i + 1)] + s)
            f = g
        return f[-1]